clients/java/src/main/java/io/zeebe/client/impl/worker/JobWorkerImpl.java

```diff
@@ -72,60 +97,99 @@ public boolean isOpen() {
     acquiringJobs.set(false);
   }
 
-  private void tryActivateJobs() {
-    final int remainingJobs = this.remainingJobs.get();
-    if (shouldActivateJobs(remainingJobs)) {
-      activateJobs();
+  /**
+   * Schedules a poll for jobs with a delay of {@code pollInterval}. Does not schedule twice if a
+   * poll is already scheduled.
+   */
+  private void schedulePoll() {
+    if (isPollScheduled.compareAndSet(false, true)) {
+      executor.schedule(this::onScheduledPoll, pollInterval, TimeUnit.MILLISECONDS);
     }
   }
 
-  private void activateJobs() {
-    final JobPoller jobPoller = this.jobPoller.getAndSet(null);
-    if (jobPoller != null) {
-      // check the condition again within the critical section
-      // to avoid race conditions that would let us exceed the buffer size
-      final int currentRemainingJobs = remainingJobs.get();
-      if (shouldActivateJobs(currentRemainingJobs)) {
-        final int maxActivatedJobs = maxJobsActive - currentRemainingJobs;
-        try {
-          jobPoller.poll(
-              maxActivatedJobs,
-              this::submitJob,
-              activatedJobs -> {
-                remainingJobs.addAndGet(activatedJobs);
-                this.jobPoller.set(jobPoller);
-              },
-              this::isOpen);
-        } catch (final Exception e) {
-          LOG.warn("Failed to activate jobs", e);
-          this.jobPoller.set(jobPoller);
-        }
-      } else {
-        this.jobPoller.set(jobPoller);
-      }
+  /** Frees up the scheduler and polls for new jobs. */
+  private void onScheduledPoll() {
+    isPollScheduled.set(false);
+    final int actualRemainingJobs = remainingJobs.get();
+    if (shouldPoll(actualRemainingJobs)) {
+      poll();
     }
   }
 
-  private boolean shouldActivateJobs(final int remainingJobs) {
+  private boolean shouldPoll(final int remainingJobs) {
     return acquiringJobs.get() && remainingJobs <= activationThreshold;
   }
 
-  private void submitJob(final ActivatedJob job) {
-    executor.execute(jobRunnableFactory.create(job, this::jobHandlerFinished));
+  private void poll() {
+    tryClaimJobPoller()
+        .ifPresent(
+            jobPoller -> {
+              // check the condition again within the critical section
+              // to avoid race conditions that would let us exceed the buffer size
+              final int actualRemainingJobs = remainingJobs.get();
+              if (shouldPoll(actualRemainingJobs)) {
+                final int maxJobsToActivate = maxJobsActive - actualRemainingJobs;
+                try {
+                  jobPoller.poll(
+                      maxJobsToActivate,
+                      this::handleJob,
+                      activatedJobs -> onPollSuccess(jobPoller, activatedJobs),
+                      error -> onPollError(jobPoller, error),
+                      this::isOpen);
+                } catch (final Exception e) {
+                  LOG.warn("Failed to activate jobs", e);
```

Comment:
Although it seems unlikely, this would also catch exceptions thrown by the `onPollSuccess` and `onPollError` (nested this includes the user implementable `BackoffSupplier`. It's being defensive here, to try to recover from the exception and allow the scheduled poll. Do we want to try to recover or should we crash immediately?

Reply:
We discussed and decided to recover (and fallback to default backoff supplier). After applying these changes, the only cases I see where this can happen is when there is a failure in `executor.execute(..)` (to handle the job or to schedule the next poll) or a direct problem with the gatewaystub itself.